packages/aft/lib/src/commands/test/test_command.dart

```diff
@@ -0,0 +1,202 @@
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:aft/aft.dart';
+import 'package:aft/src/active_device.dart';
+import 'package:aft/src/commands/test/integration_test_command.dart';
+import 'package:aft/src/commands/test/unit_test_command.dart';
+import 'package:aft/src/test_reports/print_results.dart';
+import 'package:aft/src/test_reports/test_report.dart';
+import 'package:aft/src/test_reports/test_result.dart';
+import 'package:path/path.dart' as p;
+import 'package:very_good_test_runner/very_good_test_runner.dart'
+    hide TestResult;
+import 'package:very_good_test_runner/very_good_test_runner.dart' as vgv
+    show TestResult;
+
+/// Top-level command for running Flutter, Dart and integration tests.
+class TestCommand extends AmplifyCommand {
+  TestCommand() {
+    addSubcommand(UnitTestCommand());
+    addSubcommand(IntegrationTestCommand());
+  }
+
+  @override
+  String get description =>
+      'Top-level command for running Flutter, Dart and integration tests.';
+
+  @override
+  String get name => 'test';
+}
+
+abstract class BaseTestCommand extends AmplifyCommand {
+  /// Returns the package to test, i.e. the package of the current directory.
+  Future<PackageInfo> get packageToTest async {
+    if (workingDirectory.pubspec == null) {
+      exitError('Current directory has no pubspec.yaml');
+    }
+    final packages = await allPackages;
+    return packages.values.singleWhere(
+      (el) => el.path == workingDirectory.path,
+    );
+  }
+
+  /// Gets the active devices as provided by `flutter devices`.
+  Future<List<ActiveDevice>> getActiveDevices() async {
+    final result = await executeProcess(
+      PackageFlavor.flutter,
+      const ['devices', '--machine'],
+      printStream: false,
+    );
+
+    return (json.decode(result) as List)
+        .map((i) => ActiveDevice.fromJson((i as Map).cast()))
+        .toList();
+  }
+
+  /// Executes tests for [package] and returns the list of reports.
+  Future<List<TestReport>> executeTests(
+    PackageInfo package, {
+    required List<String> arguments,
+  }) async {
+    Stream<TestEvent> testStream;
+    switch (package.flavor) {
+      case PackageFlavor.flutter:
+        testStream = flutterTest(
+          arguments: arguments,
+          workingDirectory: package.path,
+          // runInShell required for Windows
+          runInShell: true,
+        );
+        break;
+      case PackageFlavor.dart:
+        testStream = dartTest(
+          arguments: arguments,
+          workingDirectory: package.path,
+        );
+        break;
+    }
+    final reportBuilders = <int, TestReportBuilder>{};
+    final suitePaths = <int, String?>{};
+    final reports = <TestReport>[];
+    testStream = testStream.handleError((Object error) {
+      // These can be non-harmful errors such as
+      // `Warning: pubspec.yaml has overrides from pubspec_overrides.yaml`
+      // Log and evaluate if further action is needed.
+      logger.stderr(error.toString());
```

Comment:
You could fail fast in some scenarios. Although it only leads to pretty small improvement in dx and seems like these errors are always strings, so might not be worth the effort to try to differentiate between them.

For example, if the test directory doesn't exist this could throw since we know the execution cannot continue successfully. Since errors are caught and logged, the execution continues leading to another issue later (I think because the test output doesn't exist). The result is a somewhat confusing error log since it contains two errors. I had assumed it was just one error and immediately went to the stack trace, which was unhelpful.

```
Test directory "test" not found.
Bad state: No element
#0      Iterable.reduce (dart:core/iterable.dart:353:7)
#1      TestFolio.overallScore (package:aft/src/test_reports/test_folio.dart:64:26)
#2      printResults (package:aft/src/test_reports/print_results.dart:33:30)
#3      UnitTestCommand.run (package:aft/src/commands/test/unit_test_command.dart:43:5)
<asynchronous suspension>
#4      CommandRunner.runCommand (package:args/command_runner.dart:209:13)
<asynchronous suspension>
#5      main (file:///Users/nejrd/Documents/dev/amplify-flutter/packages/aft/bin/aft.dart:42:5)
<asynchronous suspension>
```

```suggestion
      if (errorString.contains('Test directory "test" not found.')) {
        throw StateError(
          'Test directory "test" not found. '
          'Are you sure you are in the correct directory?',
        );
      } else {
        // These can be non-harmful errors such as
        // `Warning: pubspec.yaml has overrides from pubspec_overrides.yaml`
        // Log and evaluate if further action is needed.
        logger.stderr(errorString);
      }
```

Reply:
Done 👍 added check before tests run:

```dart
if (packageToTest.unitTestDirectory == null) {
  exitError('No unit tests for package');
}
```